src/machine/machine_mimxrt1062_i2c.go

```diff
@@ -0,0 +1,725 @@
+	const tryMax = 0 // keep trying forever
+	rxSize := len(rxBuffer)
+	if rxSize == 0 {
+		return resultSuccess
+	}
+	// wait until there is room in the FIFO
+	result := i2c.waitForTxReady()
+	if resultSuccess != result {
+		return result
+	}
+	sizeMask := (uint32(rxSize-1) << nxp.LPI2C_MTDR_DATA_Pos) & nxp.LPI2C_MTDR_DATA_Msk
+	i2c.Bus.MTDR.Set(uint32(commandRxData) | sizeMask)
+
+	// receive data
+	for rxSize > 0 {
+		// read LPI2C receive FIFO register. the register includes a flag to
+		// indicate whether the FIFO is empty, so we can both get the data and check
+		// if we need to keep reading using a single register read.
+		var data uint32
+		try := 0
+		for 0 == tryMax || try < tryMax {
+			// check for errors on the bus
+			status := statusFlag(i2c.Bus.MSR.Get())
+			result = i2c.checkStatus(status)
+			if resultSuccess != result {
+				return result
+			}
+			// read received data, break if FIFO was non-empty
+			data = i2c.Bus.MRDR.Get()
+			if 0 == (data & nxp.LPI2C_MRDR_RXEMPTY_Msk) {
+				break
+			}
+			try++
+		}
+		// ensure we didn't timeout waiting for data
+		if 0 != tryMax && try >= tryMax {
+			return resultTimeout
+		}
+		// copy data to RX buffer
+		rxBuffer[len(rxBuffer)-rxSize] = byte(data & nxp.LPI2C_MRDR_DATA_Msk)
+		rxSize--
+	}
+	return result
+}
+
+// controllerReceive performs a polling transmit transfer on the I2C bus.
+func (i2c *I2C) controllerTransmit(txBuffer []byte) resultFlag {
+	txSize := len(txBuffer)
+	for txSize > 0 {
+		// wait until there is room in the FIFO
+		result := i2c.waitForTxReady()
+		if resultSuccess != result {
+			return result
+		}
+		// write byte into LPI2C data register
+		i2c.Bus.MTDR.Set(uint32(txBuffer[len(txBuffer)-txSize] & nxp.LPI2C_MTDR_DATA_Msk))
+		txSize--
+	}
+	return resultSuccess
+}
+
+type commandBuffer struct {
+	buffer []uint16
+	curr   uint
+	size   uint
+}
+
+func newCommandBuffer(size uint) *commandBuffer {
+	return &commandBuffer{
+		buffer: make([]uint16, size),
+		curr:   0,
+		size:   size,
+	}
+}
+
+func (cb *commandBuffer) add(cmd uint16) bool {
+	if cb.curr < cb.size {
+		cb.buffer[cb.curr] = cmd
+		cb.curr++
+		return true
+	}
+	return false
+}
+
+type transferOption struct {
+	flags          transferFlag // transfer options bit mask (0 = normal transfer)
+	peripheral     addressable  // 7-bit peripheral address
+	direction      I2CDirection // DirectionRead or DirectionWrite
+	subaddress     uint16       // peripheral sub-address (transferred MSB first)
+	subaddressSize uint16       // byte length of sub-address (maximum = 4 bytes)
+}
+
+func (i2c *I2C) controllerTransferPoll(option transferOption, data []byte) resultFlag {
+	// return an error if the bus is already in use by another controller
+	if i2c.isBusBusy() {
+		return resultBusy
+	}
+	// clear all flags
+	i2c.Bus.MSR.Set(uint32(statusClear))
+	// turn off auto-stop
+	i2c.Bus.MCFGR1.ClearBits(nxp.LPI2C_MCFGR1_AUTOSTOP)
+
+	cmd := newCommandBuffer(7)
```

Comment:
To me it looks like `commandBuffer` can easily be replaced with regular slice operations: `make([]byte, 0, 7)` and `append`. That would be shorter in code, (in my opinion) easier to understand and probably even more efficient.

Reply:
Ah, indeed. Got caught up porting C code and didn't think through the overall intent. Will remove this type.